src/testing/fast_array_utils/__init__.py

```diff
@@ -3,135 +3,42 @@
-            import cupyx.scipy.sparse as cu_sparse
-
-            return getattr(cu_sparse, cls_name)  # type: ignore[no-any-return]
-        case "dask.array", cls_name, _:
-            if TYPE_CHECKING:
-                from dask.array.core import Array as DaskArray
-            else:
-                from dask.array import Array as DaskArray
-
-            return DaskArray
-        case "h5py", "Dataset", _:
-            import h5py
-
-            return h5py.Dataset  # type: ignore[no-any-return]
-        case "zarr", "Array", _:
-            import zarr
-
-            return zarr.Array
-        case _:
-            msg = f"Unknown array class: {qualname}"
-            raise ValueError(msg)
-
-
-def random_mat(
-    shape: tuple[int, int],
-    *,
-    density: SupportsFloat = 0.01,
-    format: Literal["csr", "csc"] = "csr",  # noqa: A002
-    dtype: DTypeLike | None = None,
-    container: Literal["array", "matrix"] = "array",
-    gen: np.random.Generator | None = None,
-) -> CSBase:
-    """Create a random matrix."""
-    from scipy.sparse import random as random_spmat
-    from scipy.sparse import random_array as random_sparr
-
-    m, n = shape
-    return (
-        random_spmat(m, n, density=density, format=format, dtype=dtype, random_state=gen)
-        if container == "matrix"
-        else random_sparr(shape, density=density, format=format, dtype=dtype, random_state=gen)
-    )
-
-
-def random_array(
-    qualname: str,
-    shape: tuple[int, int],
-    *,
-    dtype: _DTypeLikeFloat32 | _DTypeLikeFloat64 | None,
-    gen: np.random.Generator | None = None,
-) -> Array:
-    """Create a random array."""
-    gen = np.random.default_rng(gen)
-
-    m = RE_ARRAY_QUAL.fullmatch(qualname)
-    assert m
-    match m["mod"], m["name"], m["inner"]:
-        case "numpy", "ndarray", None:
-            return gen.random(shape, dtype=dtype or np.float64)
-        case "scipy.sparse", (
-            "csr_array" | "csc_array" | "csr_matrix" | "csc_matrix"
-        ) as cls_name, None:
-            fmt, container = cls_name.split("_")
-            return random_mat(shape, format=fmt, container=container, dtype=dtype)  # type: ignore[arg-type]
-        case "cupy", "ndarray", None:
-            raise NotImplementedError
-        case "cupyx.scipy.sparse", ("csr_matrix" | "csc_matrix") as cls_name, None:
-            raise NotImplementedError
-        case "dask.array", cls_name, _:
-            raise NotImplementedError
-        case "h5py", "Dataset", _:
-            raise NotImplementedError
-        case "zarr", "Array", _:
-            raise NotImplementedError
-        case _:
-            msg = f"Unknown array class: {qualname}"
-            raise ValueError(msg)
+    from ._array_type import Array, ToArray  # noqa: TC004
+
+
+__all__ = [
+    "SUPPORTED_TYPES",
+    "Array",
+    "ArrayType",
+    "ConversionContext",
+    "Flags",
+    "ToArray",
+    "random_mat",
+]
+
+
+_TP_MEM = (
+    ArrayType("numpy", "ndarray", Flags.Any),
```

Comment:
Is there ever a scenario in which `Flags.Any` is not included?

Reply:
I see it is only used in a test for its own existence so not clear on its goal.

Reply:
> Is there ever a scenario in which `Flags.Any` is not included?

No, that’s it’s purpose, and that’s tested in that test.

It’s a good question: the goal is that `Flags(0)` means “nothing”. Without `Flags.Any`, `Flags(0)` would be “not (sparse or GPU or dask or disk)”, which matches `np.ndarray`! So `skip=Flags(0)` would skip `np.ndarray`. I thought about two alternatives:

1. using `None` to mean “nothing”, but I feel like `Flags(0)` feels like it should mean that.
2. adding a negative flag for everything instead (like `Flags.Dense`) but I think that makes it ambiguous how to express things: should we skip sparse or select dense?

I’m still not sure if the current one is the best approach or approach 2 from above, what do you think?

Reply:
Maybe I'm lost here but if we replace `Flags.Any` with `Flags.Dense`, apply it appropriately to the types here, and then do `Flags(0)` we would have, “not (sparse or GPU or dask or disk _or dense_)” which is everything, as desired, no? Then `~Flags(0)` would be everything.  This is very similar to two but I don't understand why we would need a negative flag here. You would not type the sparse ones as `Flag.Dense` and then you could skip `Flags.Sparse` and still get `ndarray` from `Dense`

Reply:
Yeah, except that then `Flags.Dense` would be singled out as the only flag that is also the inverse of another flag. `Any` is cleaner, but of course it’s also not maximally clean.  `~Flags(0)` already means everything!

Reply:
Moved into https://github.com/scverse/fast-array-utils/issues/34